---
date: '2022년 9월 19일 오후 8:30'
title: '비동기: 동시성 프로그래밍 1'
menu: 'javascript'
categories: ['JavaScript']
summary: 'Callback과 Promise의 함수형 프로그래밍에 대해 알아보자.'
thumbnail: './thumbnail/async_concurrency_1.png'
---
## Callback과 Promise
### Callback

함수의 인자값으로 함수를 받아 해당 함수에 인자값을 전달하면서 로직을 수행한다.

```jsx
function add10(a, callback) {
	setTimeout(() => callback(a + 10), 100);
}

add10(5, res = > {
	log(res);
});   // 15
```

### Promise

Promise객체를 만들어 return 해준다는 점이 callback함수와 가장 큰 차이이다.

```jsx
function add20(a) {
	return new Promise(resolve => setTimeout(() => resolve(a + 20), 100));
}

add20(5)
	.then(log);
```

### 차이점

1. 연속적으로 실행하는 경우 depth 차이가 발생함
2. callback은 함수를 인자로 받아 수행하는 반면, Promise는 Promise 객체를 반환한다. **로직이 수행되는 주체가 다르다.**

```jsx
add10(5, res => {
	add10(res, res => {
		add10(res, res => {
			log(res);
		});
	});
})

add20(5)
	.then(add20)
  .then(add20)
	.then(log);
```

## 비동기를 값으로 만드는 Promise

Promise와 callback의 가장 큰 차이점은 Promise는 비동기 상황을 일급 값으로 처리한다는 점이다. 대기, 성공, 실패를 다루는 일급 값으로 이루어져 있다는 점이 가장 다르고 중요하다.

Promise는 Promise로 만들어진 인스턴스를 반환하여 대기, 성공, 실패를 다루는 일급 값으로 이루어져 있는데, 이는 로직을 끝내는 것을 코드나 컨택스트로만 다루는게 아니라 대기중이라는 값을 생성한다는 점에서 Callback과 가장 큰 차이점이다.

값을 만들어서, 리턴을 하고 있기 때문에 그 이후에 떨어지는 결과로 내가 하고 싶은 일을 추가적으로 할 수 있다는 점이 가장 큰 장점이다 !!

```jsx
function add10(a, callback) {
  setTimeout(() => callback(a + 10), 100);
}

var a = add10(5, res => {
  add10(res, res => {
    add10(res, res => {
        log(res);
    });
  });
});

log(a);   // undefined

function add20(a) {
  return new Promise(resolve => setTimeout(() => resolve(a + 20), 100));
}

var b = add20(5)
  .then(add20)
  .then(add20)
  .then(log);   // Promise {<pending>}
```

- callback의 결과를 담은 a와 다르게 Promise의 반환객체를 담은 b의 출력값을 보면 Promise 객체를 보여주는데, 이는 callback에서는 반환값에 중점을 두는게 아닌 코드적인 상황(`setTimeout`)이나 Context(`Callback`)만 중점으로 둔다는 것이다.
- Promise는 즉시 Promise객체를 반환한다는 특징이 있는데, 이는 callback과는 다르게 Context 함수 로직에 이어지는 로직을 then을 통해 추가적으로 이어갈 수 있다는 것이다. 즉, Promise의 경우 비동기적으로 이루어진 상황에 대해서 값으로 다룬다는 것이고, 일급값이라는 의미가 되며 다른 곳에서 해당 일급값을 재사용할 수 있다는 의미에서 연속성을 가질 수 있다는 장점이 있다.

## 값으로서의 Promise 활용

Promise가 비동기 상황을 일급값으로 다룬다는 점을 활용한 여러가지 예제이다.

### 예제

- 아래와 같이 a라는 인자값을 f에 들어온 인자값 함수에 전달해 로직을 수행하는 go1 함수가 있다.

```jsx
const go1 = (a, f) => f(a);
const add5 = a => a + 5;

const delay100 = a = new Promise(resolve =>
	setTimeout(() => resolve(a), 100));

log(go1(10, add5));   // 15
log(go1(Promise.resolve(10), add5));   // [object Promise]5
log(go1(delay100(10), add5);   // [object Promise]5
```

go1은 a인자가 동기적으로 유효한 값인 경우에만 정상 동작하는 함수이다. a인자가 Promise객체이거나, 비동기적으로 함수의 결과가 지연되는 경우 정상적인 결과를 출력하지 못한다. 

- go1 함수에서 a인자값이 Promise인지 평가한 후 then을 수행하도록 추가한다.

```jsx
const go1 = (a, f) => a instanceof Promise ? a.then(f) : f(a);

var r2 = go1(delay100(10), add5);
r2.then(log);   // 15
```

- 비동기, 동기적으로 호출된 두 함수를 같은 형태를 가지도록 구현할 수 있다.

```jsx
const n1 = 10;
go1(go1(n1, add5), log);

const n2 = delay100(10);
go1(go1(n2, add5), log);
```

## 합성 관점에서의 Promise와 모나드

Promise는 비동기 상황에서 함수 합성을 하기 위한 모나드라고 할 수 있다. (이 얘기를 처음 들었을 때 도저히 이해하기 어려웠는데, 아래 내용들을 이해하다보면 딱 맞는 문장이라고 생각한다.)

자바스크립트는 동적타입언어이자 스크립트 언어이기 때문에 타입에 대해 엄격하지 않기에 모나드, 대수구조의 타입이 잘 붇어나지 않는 경향이 있기 때문에 자바스크립트에서는 모나드를 직접적으로 사용하거나 모나드의 개념을 이용한 사용자 정의 객체를 만들면서 구현하지는 않는다. 

그렇기 때문에 모나드라는 것을 이해하는데 너무 초점을 둘 필요는 없다.

하지만, 함수형 프로그래밍이나 함수 합성에서 모나드의 개념을 알고 있으면 좀 더 높은 퀄리티의 코드를 작성할 수 있다. 

### 모나드란?

모나드는 일종의 박스이고 박스안에 값이 들어있다고 볼수 있다. `[1]`

그리고 이 값을 이용해서 함수 합성들을 안전하게 수행해나가는 것이다. 

```jsx
const g = a => a + 1;
const f = a => a * a;

log(f(g(1)));   // 4
log(f(g()));   // NaN
```

- 인자가 없는 상태로 함수 합성 되는 경우 정상적으로 값이 출력되지 않는다.

📌  **결론**

함수합성에 사용되는 인자값이 유의미한 값이 아닌 경우 문제가 발생한다는 것인데, 실무에서는 인자값으로 어떤 값이 올 지 모르고 빈값이 올 가능성도 충분하기 때문에 위의 함수합성은 안전하지 않은 함수 합성이다. 이처럼 인자값이 어떤 값이 올 지 모르는 불안한 상황에서 함수 합성을 어떻게 안전하게 할 지 고려하여 나온 것이 **모나드**이다.

### 예제

아래와 같이 박스를 가지고 있고, 그 내부에 실제 효과나 연산에 필요한 값을 가지고 있고 이를 통해 함수 합성을 한다. 반환 타입이 Array인걸 볼 수 있는데 타입이 중요한 것은 아니므로 꺼내어주는 것이 좋다.

```jsx
log([1].map(g).map(f));   // [4]
[1].map(g).map(f).forEach(a => log(a)); // 4

[].map(g).map(f).forEach(a => log(a)); // 결과 없음
```

- 안에 아무런 값이 없어도 어떤 잘못된 출력이나 에러가 발생하지 않는다. 실제 최종 결과 값을 도출해내는 `forEach` 함수에 도착하기전에 안전하게 종료된다.

> **모나드**:  x라는 인자값을 a라는 함수에게 전달했을때 a라는 함수가 수행 된 이후 그 값이 바로 b라는 함수에 전달되는 것을 함수 합성이라 하는데, 이러한 함수 합성을 안전하게 수행하도록 하는 것을 모나드라고 하고, 대표적으로 Promise가 있다.
⇒ b(a(x))
> 

### Promise는?

```jsx
Promise.resolve(1).then(g).then(f).then(log);   // 4
```

Promise는 비동기적으로 일어나는 상황에 대해 안전하게 함수 합성을 하기 위한 도구이다. 지연되어 나오는 값들에 대해서 안전하게 핸들링 하기 위한 도구라고 할 수 있다.

## Kleisli Composition 관점에서의 Promise

Promise는 Kleisli Composition을 지원하는 도구라고 볼 수 있는데, Kleisli Composition 혹은 Kleisli Arrow라고 불리는 함수 합성 방법은 오류가 있을수 있는 상황에서의 함수합성을 안전하게 하는 하나의 규칙이라고 볼 수 있다.

Kleisli Composition은 들어오는 인자가 잘못되어서 함수에서 오류가 나는 상황이나 정확한 인자가 들어왔더라도 어떤 함수가 의존하고 있는 외부의 상태에 의해서 결과를 전달 할 수 없을 때 이것을 해결하기 위한 함수 합성이라고 볼 수 있다.

### 예제

- 유저목록에서 id로 해당 유저를 찾아 이름을 반환하는 로직
    
    ```jsx
    var users = [
    	{ id: 1, name: 'aa' },
    	{ id: 2, name: 'bb' },
    	{ id: 3, name: 'cc' }
    ];
    
    const getUserById = id =>	find(u => u.id === id, users);
    
    const f = ({name}) => name;
    const g = getUserById;
    const fg = id => f(g(id));
    
    const r = fg(2);
    log(r);   // bb
    ```
    
    - users에서 인자값으로 받은 id와 동일한 user를 찾는 함수 getuserById
    - name을 구조분해하여 얻어 반환하는 f
    - getuserById를 값으로 취급하는 g
    - f와 g를 합성해 users에서 특정 id의 name을 추출해 반환하는 fg
    - fg(2);를 호출하면 g(2)가 실행되어 users에서 id가 2인 obj를 꺼내전달하고 ({id:2, name:'bb})
    f함수에서는 name을 구조분해하여 추출후 반환한다.

여기서 찾고자 하는 id가 현재 user에 없는 4, 5, 6이거나 기존 id들이 외부 상황에 의해 변한다면 오류가 발생한다.

```jsx
const r = fg(5); //Uncaught TypeError: Cannot destructure property 'name' of 'undefined' as it is undefined.

users.pop();
users.pop();
const r = fg(2)//Uncaught TypeError: Cannot destructure property 'name' of 'undefined' as it is undefined.
```

잘못된 값 혹은 외부에 변화에 따라 에러가 발생할 수 있는데, 이러한 상황에서 문제가 발생하지 않도록 하는 것을 **Kleisli Arrow**라고 한다.

```jsx
const getUserById = id => find(u => u.id === id, users)|| Promise.reject("없어요!");
const fg = id => Promise.resolve(id).then(g).then(f).catch(a=>a);
users.pop();
users.pop();
fg(2).then(log); //없어요!
```

- `getUserById`의 결과가 없는 경우 `Promise`객체를 반환하게 되고 `fg`에서는 해당객체의 `reject`가 호출되면 `catch`부분에서 받아서 출력함으로써 에러를 발생하지 않도록 한다.

## go, pipe, reduce에서 비동기 제어

비동기를 다루는 Promise객체를 이용해 go, pipe, reduce함수들에서도 비동기를 값으로 다루는 성직을 이용하여 Promise와 같이 비동기 상황에 놓여져도 잘 대응하는 함수를 만들 수 있다. 또는 Kleisli처럼 중간에 reject가 발생했을 경우에도 대응하는 방법을 적용할 수 있다.

### 예제

```jsx
go(1,
    a => a + 10,
		a => Promise.resolve(a + 100),
    a => a + 1000,
    log
);   // [object Promise]1000
```

위와 같이 `a + 100` 이 Promise타입으로 바뀌는 경우에 올바르지 않은 값이 출력된다.

이것은 현재 go함수를 통해 전달되어 실행되는 함수가 reduce인데 여기서 acc가 Promise가 되면서 (`f(Promise, a)` ) 다음 루프를 도는 경우 에러가 발생한다.

```jsx
const go = (...args) => reduce((a, f) => f(a), args);
const reduce = curry((f, acc, iter) => {
    if (!iter) {
        iter = acc[Symbol.iterator]();
        acc = iter.next().value;
    }
    for (const a of iter) {
        acc = f(acc, a);
    }
    return acc;
});
```

### 해결 방법 - 1

```jsx
const reduce = curry((f, acc, iter) => {
    if (!iter) {
        iter = acc[Symbol.iterator]();
        acc = iter.next().value;
    }
    for (const a of iter) {
        acc = acc instanceof Promise ? acc.then(acc => f(acc,a)) : f(acc, a);
    }
    return acc;
}); // 1111
```

Promise인 경우 then을 실행해주는 방법으로 해결해 줄 수 있다. 하지만 Promise를 만나게 되면 그 뒤로는 계속해서 Promise Chaning이 걸리게 되고, 이런 함수 합성이 많은 경우 불 필요한 분기를 타게 되면서 성능 저하가 일어날 수 있다.

### 해결 방법 - 2

```jsx
const reduce = curry((f, acc, iter) => {
    if (!iter) {
        iter = acc[Symbol.iterator]();
        acc = iter.next().value;
    }
    return function recur(acc) {
        for (const a of iter) {
            acc = f(acc, a);
            if (acc instanceof Promise) return acc.then(recur);
        }
        return acc;
    }(acc);
});

go(1,
    a => a + 10,
    a => Promise.resolve(a + 100),
    a => a + 1000,
    log
);   //1111
```

재귀를 이용한 방법을 사용한다. 유명함수를 이용해 return값에 `recur`함수를 선언해서 내부에서 우선 인자값 함수를 실행한 후 해당 값이 Promise이면 `acc.then(recure)`로 재귀호출을 하여 Promise의 실제 결과 값이 인자로 전달되는 방식이다.

### 보강 - 첫 인자값이 Promise인 경우

go에서 사용되는 첫 번째 인자값이 Promise인 경우에는 아직 에러가 발생한다.

```jsx
go(Promise.resolve(1),
	a => a + 10,
	a => Promise.resolve(a + 100),
	a => a + 1000,
	log
);
```

위와 같이 첫번째 인자부터 Promise인 경우 reduce의 유명함수 실행시 첫번째 인자를 Promise로 전달하기 때문에 `f(acc, a)` 에서 오류가 발생한다. 이런 경우 첫번째 인자값이 Promise인 경우 then을 실행하는 `go1` 함수를 이용하여 처리를 해주어야한다.

```jsx
const go1 = (a, f) => a instanceof Promise ? a.then(f) : f(a);

const reduce = curry((f, acc, iter) => {
    if (!iter) {
        iter = acc[Symbol.iterator]();
        acc = iter.next().value;
    }
    return go1(acc, function recur(acc) {
        for (const a of iter) {
            acc = f(acc, a);
            if(acc instanceof Promise) return acc.then(recur);
        }
        return acc;
    });
});

go(Promise.resolve(1),
    a=>a+10,
    a=>Promise.resolve(a+100),
    a=>a+1000,
    log
);//1111
```

## Promise.then의 중요한 규칙

1. 여러개의 `resolve`가 중첩되어 있어도 한번의 `then`으로 꺼내어 볼 수 있다.
2. 아무리 `Promise`가 여러개로 중첩되어있어도 가장 내부에 있는 결과값이 도출되어 전달된다.

```jsx
Promise.resolve(Promise.resolve(1)).then(function (a) {
  log(a);
});

new Promise(resolve => resolve(new Promise(resolve => resolve(1)))).then(log);
```